### PR TITLE
feat(k8s): cyclonedx kbom support

### LINKED ISSUE #4029: kubernetes bill of materials support
it is required to produce kubernetes (core components) bill of materials supported:
- kind
- minikube
- rancher
- openshift
- gke
- eks
- aks

### PR DESCRIPTION
**Pull request by chen-keinan**
## Description
support cyclonedx kubernetes bom

## Related issues
- Close #4029

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).

usage:

```console
# trivy k8s cluster --format cyclonedx
```

```json
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:280a88b1-3f06-458e-853a-b859a6b76ed3",
  "version": 1,
  "metadata": {
    "timestamp": "2023-06-15T06:35:12+00:00",
    "tools": [
      {
        "vendor": "aquasecurity",
        "name": "trivy",
        "version": "dev"
      }
    ],
    "component": {
      "bom-ref": "3ca4755b-dcab-49c7-8593-a62e132cfa18",
      "type": "container",
      "name": "kind-kind",
      "properties": []
    }
  },
  "components": [
    {
      "bom-ref": "09deecd9-4d71-4d41-bb97-381f41461743",
      "type": "operating-system",
      "name": "ubuntu",
      "version": "21.04",
      "properties": [
        {
          "name": "aquasecurity:trivy:Class",
          "value": "os-pkgs"
        },
        {
          "name": "aquasecurity:trivy:Type",
          "value": "ubuntu"
        }
      ]
    },
    {
      "bom-ref": "0ed5877d-de56-458a-b596-973397997b41",
      "type": "application",
      "name": "coredns-558bd4d5db-t5f4l",
      "properties": [
        {
          "name": "aquasecurity:trivy:addons",
          "value": "kube-dns"
        }
      ]
    },
    {
      "bom-ref": "40de83f5-a856-4104-91a0-1d59fcc58803",
      "type": "application",
      "name": "kube-controller-manager-kind-control-plane",
      "properties": [
        {
          "name": "aquasecurity:trivy:control_plane_components",
          "value": "kube-controller-manager"
        }
      ]
    },
    {
      "bom-ref": "497c7cad-12cc-4a8e-917e-fe6b9702db42",
      "type": "application",
      "name": "kube-apiserver-kind-control-plane",
      "properties": [
        {
          "name": "aquasecurity:trivy:control_plane_components",
          "value": "kube-apiserver"
        }
      ]
    },
    {
      "bom-ref": "4df8d851-8cc3-4d64-8a7e-6ed1efb3f97d",
      "type": "application",
      "name": "node-core-components",
      "properties": [
        {
          "name": "aquasecurity:trivy:Class",
          "value": "lang-pkgs"
        },
        {
          "name": "aquasecurity:trivy:Type",
          "value": "golang"
        }
      ]
    },
    {
      "bom-ref": "4fbc3673-54fc-4897-b688-47b578c9cfd0",
      "type": "application",
      "name": "kube-proxy-hgmbt",
      "properties": [
        {
          "name": "aquasecurity:trivy:addons",
          "value": "kube-proxy"
        }
      ]
    },
    {
      "bom-ref": "603e1d1e-98cf-42ba-920e-563a5cde410e",
      "type": "application",
      "name": "kindnet-rt7rc",
      "properties": [
        {
          "name": "aquasecurity:trivy:addons",
          "value": "kindnet"
        }
      ]
    },
    {
      "bom-ref": "7ea39473-d30b-4a0c-bea7-ef0280b228bb",
      "type": "application",
      "name": "etcd-kind-control-plane",
      "properties": [
        {
          "name": "aquasecurity:trivy:control_plane_components",
          "value": "etcd"
        }
      ]
    },
    {
      "bom-ref": "9aeb22d9-309b-4553-b42d-8aa5f08ac646",
      "type": "container",
      "name": "kind-control-plane",
      "properties": [
        {
          "name": "aquasecurity:trivy:architecture",
          "value": "arm64"
        },
        {
          "name": "aquasecurity:trivy:host_name",
          "value": "kind-control-plane"
        },
        {
          "name": "aquasecurity:trivy:kernel_version",
          "value": "6.2.15-300.fc38.aarch64"
        },
        {
          "name": "aquasecurity:trivy:node_role",
          "value": "master"
        },
        {
          "name": "aquasecurity:trivy:operating_system",
          "value": "linux"
        }
      ]
    },
    {
      "bom-ref": "da0371db-269b-45d6-b16f-18125d6cc221",
      "type": "application",
      "name": "kube-scheduler-kind-control-plane",
      "properties": [
        {
          "name": "aquasecurity:trivy:control_plane_components",
          "value": "kube-scheduler"
        }
      ]
    },
    {
      "bom-ref": "fc2c920d-62c9-4265-9696-09ca8f243f5f",
      "type": "application",
      "name": "coredns-558bd4d5db-gmw9d",
      "properties": [
        {
          "name": "aquasecurity:trivy:addons",
          "value": "kube-dns"
        }
      ]
    },
    {
      "bom-ref": "pkg:golang/github.com%2Fcontainerd%2Fcontainerd@1.5.2",
      "type": "library",
      "name": "github.com/containerd/containerd",
      "version": "1.5.2",
      "purl": "pkg:golang/github.com%2Fcontainerd%2Fcontainerd@1.5.2",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "golang"
        }
      ]
    },
    {
      "bom-ref": "pkg:golang/k8s.io%2Fkubelet@1.21.1",
      "type": "library",
      "name": "k8s.io/kubelet",
      "version": "1.21.1",
      "purl": "pkg:golang/k8s.io%2Fkubelet@1.21.1",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "golang"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/coredns@sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8?repository_url=k8s.gcr.io%2Fcoredns%2Fcoredns\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/coredns/coredns",
      "version": "sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8",
      "purl": "pkg:oci/coredns@sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8?repository_url=k8s.gcr.io%2Fcoredns%2Fcoredns\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/coredns/coredns:1.8.0"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/etcd@sha256:05b738aa1bc6355db8a2ee8639f3631b908286e43f584a3d2ee0c472de033c28?repository_url=k8s.gcr.io%2Fetcd\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/etcd",
      "version": "sha256:05b738aa1bc6355db8a2ee8639f3631b908286e43f584a3d2ee0c472de033c28",
      "purl": "pkg:oci/etcd@sha256:05b738aa1bc6355db8a2ee8639f3631b908286e43f584a3d2ee0c472de033c28?repository_url=k8s.gcr.io%2Fetcd\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/etcd:3.4.13-0"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/kindnetd@sha256:f37b7c809e5dcc2090371f933f7acb726bb1bffd5652980d2e1d7e2eff5cd301?repository_url=index.docker.io%2Fkindest%2Fkindnetd\u0026arch=",
      "type": "container",
      "name": "index.docker.io/kindest/kindnetd",
      "version": "sha256:f37b7c809e5dcc2090371f933f7acb726bb1bffd5652980d2e1d7e2eff5cd301",
      "purl": "pkg:oci/kindnetd@sha256:f37b7c809e5dcc2090371f933f7acb726bb1bffd5652980d2e1d7e2eff5cd301?repository_url=index.docker.io%2Fkindest%2Fkindnetd\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "index.docker.io/kindest/kindnetd:20210326-1e038dc5"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/kube-apiserver@sha256:18e61c783b41758dd391ab901366ec3546b26fae00eef7e223d1f94da808e02f?repository_url=k8s.gcr.io%2Fkube-apiserver\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/kube-apiserver",
      "version": "sha256:18e61c783b41758dd391ab901366ec3546b26fae00eef7e223d1f94da808e02f",
      "purl": "pkg:oci/kube-apiserver@sha256:18e61c783b41758dd391ab901366ec3546b26fae00eef7e223d1f94da808e02f?repository_url=k8s.gcr.io%2Fkube-apiserver\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/kube-apiserver:1.21.1"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/kube-controller-manager@sha256:0c6dccae49de8003ee4fa06db04a9f13bb46cbaad03977e6baa21174f2dba2fc?repository_url=k8s.gcr.io%2Fkube-controller-manager\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/kube-controller-manager",
      "version": "sha256:0c6dccae49de8003ee4fa06db04a9f13bb46cbaad03977e6baa21174f2dba2fc",
      "purl": "pkg:oci/kube-controller-manager@sha256:0c6dccae49de8003ee4fa06db04a9f13bb46cbaad03977e6baa21174f2dba2fc?repository_url=k8s.gcr.io%2Fkube-controller-manager\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/kube-controller-manager:1.21.1"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/kube-proxy@sha256:4bbef4ca108cdc3b99fe23d487fa4fca933a62c4fc720626a3706df9cef63b21?repository_url=k8s.gcr.io%2Fkube-proxy\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/kube-proxy",
      "version": "sha256:4bbef4ca108cdc3b99fe23d487fa4fca933a62c4fc720626a3706df9cef63b21",
      "purl": "pkg:oci/kube-proxy@sha256:4bbef4ca108cdc3b99fe23d487fa4fca933a62c4fc720626a3706df9cef63b21?repository_url=k8s.gcr.io%2Fkube-proxy\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/kube-proxy:1.21.1"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    },
    {
      "bom-ref": "pkg:oci/kube-scheduler@sha256:8c783dd2520887cc8e7908489ffc9f356c82436ba0411d554237a0b9632c9b87?repository_url=k8s.gcr.io%2Fkube-scheduler\u0026arch=",
      "type": "container",
      "name": "k8s.gcr.io/kube-scheduler",
      "version": "sha256:8c783dd2520887cc8e7908489ffc9f356c82436ba0411d554237a0b9632c9b87",
      "purl": "pkg:oci/kube-scheduler@sha256:8c783dd2520887cc8e7908489ffc9f356c82436ba0411d554237a0b9632c9b87?repository_url=k8s.gcr.io%2Fkube-scheduler\u0026arch=",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgID",
          "value": "k8s.gcr.io/kube-scheduler:1.21.1"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "oci"
        }
      ]
    }
  ],
  "dependencies": [
    {
      "ref": "09deecd9-4d71-4d41-bb97-381f41461743",
      "dependsOn": []
    },
    {
      "ref": "0ed5877d-de56-458a-b596-973397997b41",
      "dependsOn": [
        "pkg:oci/coredns@sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8?repository_url=k8s.gcr.io%2Fcoredns%2Fcoredns\u0026arch="
      ]
    },
    {
      "ref": "3ca4755b-dcab-49c7-8593-a62e132cfa18",
      "dependsOn": [
        "0ed5877d-de56-458a-b596-973397997b41",
        "40de83f5-a856-4104-91a0-1d59fcc58803",
        "497c7cad-12cc-4a8e-917e-fe6b9702db42",
        "4fbc3673-54fc-4897-b688-47b578c9cfd0",
        "603e1d1e-98cf-42ba-920e-563a5cde410e",
        "7ea39473-d30b-4a0c-bea7-ef0280b228bb",
        "9aeb22d9-309b-4553-b42d-8aa5f08ac646",
        "da0371db-269b-45d6-b16f-18125d6cc221",
        "fc2c920d-62c9-4265-9696-09ca8f243f5f"
      ]
    },
    {
      "ref": "40de83f5-a856-4104-91a0-1d59fcc58803",
      "dependsOn": [
        "pkg:oci/kube-controller-manager@sha256:0c6dccae49de8003ee4fa06db04a9f13bb46cbaad03977e6baa21174f2dba2fc?repository_url=k8s.gcr.io%2Fkube-controller-manager\u0026arch="
      ]
    },
    {
      "ref": "497c7cad-12cc-4a8e-917e-fe6b9702db42",
      "dependsOn": [
        "pkg:oci/kube-apiserver@sha256:18e61c783b41758dd391ab901366ec3546b26fae00eef7e223d1f94da808e02f?repository_url=k8s.gcr.io%2Fkube-apiserver\u0026arch="
      ]
    },
    {
      "ref": "4df8d851-8cc3-4d64-8a7e-6ed1efb3f97d",
      "dependsOn": [
        "pkg:golang/github.com%2Fcontainerd%2Fcontainerd@1.5.2",
        "pkg:golang/k8s.io%2Fkubelet@1.21.1"
      ]
    },
    {
      "ref": "4fbc3673-54fc-4897-b688-47b578c9cfd0",
      "dependsOn": [
        "pkg:oci/kube-proxy@sha256:4bbef4ca108cdc3b99fe23d487fa4fca933a62c4fc720626a3706df9cef63b21?repository_url=k8s.gcr.io%2Fkube-proxy\u0026arch="
      ]
    },
    {
      "ref": "603e1d1e-98cf-42ba-920e-563a5cde410e",
      "dependsOn": [
        "pkg:oci/kindnetd@sha256:f37b7c809e5dcc2090371f933f7acb726bb1bffd5652980d2e1d7e2eff5cd301?repository_url=index.docker.io%2Fkindest%2Fkindnetd\u0026arch="
      ]
    },
    {
      "ref": "7ea39473-d30b-4a0c-bea7-ef0280b228bb",
      "dependsOn": [
        "pkg:oci/etcd@sha256:05b738aa1bc6355db8a2ee8639f3631b908286e43f584a3d2ee0c472de033c28?repository_url=k8s.gcr.io%2Fetcd\u0026arch="
      ]
    },
    {
      "ref": "9aeb22d9-309b-4553-b42d-8aa5f08ac646",
      "dependsOn": [
        "09deecd9-4d71-4d41-bb97-381f41461743",
        "4df8d851-8cc3-4d64-8a7e-6ed1efb3f97d"
      ]
    },
    {
      "ref": "da0371db-269b-45d6-b16f-18125d6cc221",
      "dependsOn": [
        "pkg:oci/kube-scheduler@sha256:8c783dd2520887cc8e7908489ffc9f356c82436ba0411d554237a0b9632c9b87?repository_url=k8s.gcr.io%2Fkube-scheduler\u0026arch="
      ]
    },
    {
      "ref": "fc2c920d-62c9-4265-9696-09ca8f243f5f",
      "dependsOn": [
        "pkg:oci/coredns@sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8?repository_url=k8s.gcr.io%2Fcoredns%2Fcoredns\u0026arch="
      ]
    },
    {
      "ref": "pkg:golang/github.com%2Fcontainerd%2Fcontainerd@1.5.2",
      "dependsOn": []
    },
    {
      "ref": "pkg:golang/k8s.io%2Fkubelet@1.21.1",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/coredns@sha256:1a1f05a2cd7c2fbfa7b45b21128c8a4880c003ca482460081dc12d76bfa863e8?repository_url=k8s.gcr.io%2Fcoredns%2Fcoredns\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/etcd@sha256:05b738aa1bc6355db8a2ee8639f3631b908286e43f584a3d2ee0c472de033c28?repository_url=k8s.gcr.io%2Fetcd\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/kindnetd@sha256:f37b7c809e5dcc2090371f933f7acb726bb1bffd5652980d2e1d7e2eff5cd301?repository_url=index.docker.io%2Fkindest%2Fkindnetd\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/kube-apiserver@sha256:18e61c783b41758dd391ab901366ec3546b26fae00eef7e223d1f94da808e02f?repository_url=k8s.gcr.io%2Fkube-apiserver\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/kube-controller-manager@sha256:0c6dccae49de8003ee4fa06db04a9f13bb46cbaad03977e6baa21174f2dba2fc?repository_url=k8s.gcr.io%2Fkube-controller-manager\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/kube-proxy@sha256:4bbef4ca108cdc3b99fe23d487fa4fca933a62c4fc720626a3706df9cef63b21?repository_url=k8s.gcr.io%2Fkube-proxy\u0026arch=",
      "dependsOn": []
    },
    {
      "ref": "pkg:oci/kube-scheduler@sha256:8c783dd2520887cc8e7908489ffc9f356c82436ba0411d554237a0b9632c9b87?repository_url=k8s.gcr.io%2Fkube-scheduler\u0026arch=",
      "dependsOn": []
    }
  ],
  "vulnerabilities": []
}

```

### CODE REVIEW

**Review by AnaisUrlichs — APPROVED** [review]

File: docs/docs/target/kubernetes.md [hunk 1]
```diff
@@ -343,3 +343,12 @@ Trivy has a native [Kubernetes Operator][operator] which continuously scans your
...
 [trivy-operator]: https://aquasecurity.github.io/trivy-operator/latest
+
+## SBOM
+
+Trivy supports the generation of Kubernetes Bill of Materials (KBOM) for kubernetes cluster control plane components, node components and addons.
+
+### Generation
+
+Trivy can generate KBOM in cyclonedx format for kubernetes cluster core components .
+See [here](../supply-chain/sbom.md) for the detail.
```
Comment (itaysk) [inline docs/docs/target/kubernetes.md:354]:
```suggestion
## KBOM
KBOM, Kubernetes Bill of Materials, is a manifest of all the important components that make up your Kubernetes cluster – Control plane components, Node Components, and Addons, including their versions and images. Which “api-server” version are you currently running? Which flavor of “kubelet” is running on each node? What kind of etcd or storage are you currently using? And most importantly – are there any vulnerabilities known to affect these components? These are all questions that KBOM can help you answer.

Trivy can generate KBOM in CycloneDX format:

```sh
trivy k8s cluster --format cyclonedx
```

```

File: docs/docs/supply-chain/sbom.md [hunk 2]
```diff
@@ -19,6 +19,13 @@ $ trivy image --format spdx-json --output result.json alpine:3.15
 $ trivy fs --format cyclonedx --output result.json /app/myproject
 ```
 
+Kubernetes sbom i.e. kbom
+Kubernetes sbom referred to as kbom, produces the k8s core components (Control Plane Components, Node Components and Addons) bill of material.
+
+```
+trivy k8s cluster --format cyclonedx --output result.json
+```
+
```
Comment (itaysk) [inline docs/docs/supply-chain/sbom.md:28]: I think this is not needed here if you accept the other suggestion

File: docs/docs/target/kubernetes.md (hunk 1 above)
Comment (itaysk) [reply docs/docs/target/kubernetes.md:354]: needs some more documentation if we have it: examples, supported platforms, limitation, etc.

**Comment by chen-keinan** [comment]
@itaysk take a look at the `core component` properties, the above example is updated following to code changes

File: pkg/sbom/cyclonedx/marshal.go [hunk 3]
```diff
@@ -59,23 +60,27 @@ func NewMarshaler(version string, opts ...core.Option) *Marshaler {
...
+	root, err := MarshalReport(report)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to marshal report: %w", err)
 	}
 
 	return e.core.Marshal(root), nil
 }
 
-func (e *Marshaler) MarshalReport(r types.Report) (*core.Component, error) {
+func (e *Marshaler) MarshalKbom(report k8s.Report) (*cdx.BOM, error) {
```
Comment (knqyf263) [inline pkg/sbom/cyclonedx/marshal.go:71]: I don't want to add MarshalKbom here. Kubernetes should directly call core.Marshal.

File: pkg/report/cyclonedx/cyclonedx.go [hunk 4]
```diff
@@ -31,12 +32,22 @@ func (w Writer) Write(report types.Report) error {
 	if err != nil {
 		return xerrors.Errorf("CycloneDX marshal error: %w", err)
 	}
+	return w.encodeBom(bom)
+}
 
+func (w Writer) WriteKbom(report rep.Report) error {
```
Comment (knqyf263) [inline pkg/report/cyclonedx/cyclonedx.go:38]: It is kubernetes-specific. I think it should be located under `pkg/k8s/report`.

Comment (knqyf263) [inline pkg/report/cyclonedx/cyclonedx.go:38]: It is kubernetes-specific. I think it should be located under `pkg/k8s/report`.

File: pkg/k8s/scanner/scanner.go [hunk 5]
```diff
@@ -143,3 +176,140 @@ func (s *Scanner) filter(ctx context.Context, r types.Report, artifact *artifact
...
+				Results: types.Results{
+					{
+						Target:   "containers",
+						Type:     "oci",
+						Class:    types.ClassK8sComponents,
+						Packages: packages,
+					},
+				},
+			}
+			podComp, err := cyclonedx.MarshalReport(podReport)
```
Comment (knqyf263) [inline pkg/k8s/scanner/scanner.go:218]: I don't think we need to call `MarshalReport` here. You can directly construct core.Component.

```
podComp := core.Component {
	Type: cdx.Container,
	Name: comp.Name,
        ...
	Properties: comp.Properties,
}
```

File: pkg/report/cyclonedx/cyclonedx.go (hunk 4 above)
Comment (chen-keinan) [reply pkg/report/cyclonedx/cyclonedx.go:38]: we might need to add this function instead :
```
func (w Writer) WriteComponent(component core.Component) {
	bom:=w.marshaler.core.Marshal(component)
	return encodeBom(bom)
}
```
the other option is to expose `marshaler` and `Marshaler.core`
what do you prefer?

File: pkg/sbom/cyclonedx/marshal.go (hunk 3 above)
Comment (chen-keinan) [reply pkg/sbom/cyclonedx/marshal.go:71]: we might need to add this function :
```
func (e *Marshaler) MarshalComponent(component *core.Component) (*cdx.BOM, error) {
	return e.core.Marshal(component), nil
}
```
the other option is to expose  `Marshaler.core`

what do you prefer?

File: pkg/k8s/scanner/scanner.go [hunk 6]
```diff
@@ -143,3 +170,167 @@ func (s *Scanner) filter(ctx context.Context, r types.Report, artifact *artifact
...
+				name := fmt.Sprintf("%s/%s", c.Registry, c.Repository)
+				cDigest := c.Digest
+				if strings.Index(c.Digest, string(digest.SHA256)) == -1 {
+					cDigest = fmt.Sprintf("%s:%s", string(digest.SHA256), cDigest)
+				}
+				imageDigest, err := cn.NewDigest(fmt.Sprintf("%s@%s", name, cDigest))
+				if err != nil {
+					return nil, xerrors.Errorf("failed to parse digest: %w", err)
+				}
+				imagePurl := purl.PackageURL{PackageURL: purl.OciPurl(imageDigest, "")}
```
Comment (knqyf263) [inline pkg/k8s/scanner/scanner.go:204]: I think we can leverage the existing function.

```suggestion
				imagePURL, err := purl.NewPackageURL(purl.TypeOCI, types.Metadata{
					RepoDigests: []string{fmt.Sprintf("%s@%s", name, cDigest)},
				}, ftypes.Package{})
				if err != nil {
					return nil, xerrors.Errorf("failed to create PURL: %w", err)
				}
```

File: pkg/k8s/scanner/scanner_test.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+type coreComponents []*core.Component
+
+func (a coreComponents) Len() int           { return len(a) }
+func (a coreComponents) Less(i, j int) bool { return a[i].Name < a[j].Name }
+func (a coreComponents) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func sortNodeComponents(component *core.Component) {
+	nodeComp := findComponentByName(component, "node-core-components")
+	sort.Sort(coreComponents(nodeComp.Components))
```
Comment (knqyf263) [inline pkg/k8s/scanner/scanner_test.go:201]: If you just want to sort by name, you don't need the new type implementing `sort.Interface`.

```suggestion
	sort.Slice(nodeComp.Components, func(i, j int) bool {
		return nodeComp.Components[i].Name < nodeComp.Components[j].Name
	})
```

File: pkg/report/cyclonedx/cyclonedx.go [hunk 8]
```diff
@@ -31,12 +32,22 @@ func (w Writer) Write(report types.Report) error {
 	if err != nil {
 		return xerrors.Errorf("CycloneDX marshal error: %w", err)
 	}
+	return w.encodeBom(bom)
+}
 
+func (w Writer) WriteCoreComponents(component *core.Component) error {
```
Comment (knqyf263) [inline pkg/report/cyclonedx/cyclonedx.go:38]: I aim to keep `pkg/report` and `pkg/k8s/report` independent. Can we add a tiny writer in `pkg/k8s/report/cyclonedx.go`?

```
// Writer implements types.Writer
type Writer struct {
	encoder cdx.BOMEncoder
	marshaler *cyclonedx.Marshaler
}

func NewWriter(output io.Writer, appVersion string) Writer {
	encoder := cdx.NewBOMEncoder(w.output, w.format)
	encoder.SetPretty(true)
	return Writer {
		encoder: encoder,
		marshaler: core.NewCycloneDX(appVersion),
	}
}

func (w Writer) Write(component *core.Component) error {
	bom, err := w.marshaler.Marshal(report)
	if err != nil {
		return xerrors.Errorf("CycloneDX marshal error: %w", err)
	}
	return w.encoder.Encode(bom)
}
```

File: pkg/k8s/scanner/scanner.go [hunk 9]
```diff
@@ -143,3 +172,160 @@ func (s *Scanner) filter(ctx context.Context, r types.Report, artifact *artifact
...
+				},
+				Components: []*core.Component{
+					{
+						Type:    cdx.ComponentTypeLibrary,
+						Name:    kubelet,
+						Version: kubeletVersion,
+						Properties: map[string]string{
+							cyc.PropertyPkgType: golang,
+						},
+						PackageURL: &purl.PackageURL{
```
Comment (knqyf263) [inline pkg/k8s/scanner/scanner.go:312]: For PURL, we should use canonical names, `k8s.io/kubelet`.
https://github.com/kubernetes/kubelet/blob/e70f2e547c08910ba3c82c34dbfa0108cb1631eb/go.mod#L3

Comment (knqyf263) [reply pkg/k8s/scanner/scanner.go:312]: Other components as well, like `k8s.io/apiserver`. You can see the list here.
https://github.com/kubernetes/kubernetes/tree/e56002ab0495de16a5c55463c4591f9c4320705b/staging

**Comment by chen-keinan** [comment]
@knqyf263 comments has been addressed

File: go.mod [hunk 10]
```diff
@@ -392,13 +392,17 @@ require (
...
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
 // v1.2.0 is taken from github.com/open-policy-agent/opa v0.42.0
 // v1.2.0 incompatible with github.com/docker/docker v23.0.0-rc.1+incompatible
 replace oras.land/oras-go => oras.land/oras-go v1.1.1
+
+replace github.com/google/go-containerregistry => github.com/google/go-containerregistry v0.14.0
+
+replace k8s.io/kubectl => k8s.io/kubectl v0.26.3
```
[the remark this answers is not in the document]
Comment (chen-keinan) [reply go.mod:406]: sure, created [PR](https://github.com/aquasecurity/defsec/pull/1344) for defsec

[the remark this answers is not in the document]
Comment (chen-keinan) [reply go.mod]: @knqyf263 I have replaced `defsec` with `helm v3.12.1` still getting error as result from dependency issue , see my comments on `go.mod` file, could be replacing kubectl is the best option until latest `github.com/docker/docker` can be used, wdyt?

[the remark this answers is not in the document]
Comment (knqyf263) [reply go.mod:405]: Replacing `oras-go` helps. It works in my local env.
https://github.com/helm/helm/issues/11821#issuecomment-1587424133

**Comment by knqyf263** [comment]
The database is downloaded with KBOM. Is it intended?

```
$ trivy k8s cluster --format cyclonedx
2023-06-19T11:14:06.338+0300    INFO    Loaded trivy.yaml
2023-06-19T11:14:06.460+0300    INFO    Need to update DB
2023-06-19T11:14:06.460+0300    INFO    DB Repository: ghcr.io/aquasecurity/trivy-db
2023-06-19T11:14:06.460+0300    INFO    Downloading DB...
13.66 MiB / 37.76 MiB [----------------------------->____________________________________________________] 36.16% 552.13 KiB p/s ETA 44s
```

**Comment by knqyf263** [comment]
>"name": "aquasecurity:trivy:control_plane_components",

Can we use the upper camel case to get aligned with other properties?

>"name": "aquasecurity:trivy:PkgID",

**Review by knqyf263 — APPROVED** [review]
LGTM🚢

**Pull request merged**